text/0000-decorators.md

```diff
@@ -37,158 +113,768 @@ export default class Profile extends Component {
 }
 ```
 
-Decorators bring a more natural way of declaring computed properties to native classes than the current computed macros.
+Native decorators bring a more natural way of declaring computed properties to native classes than the current computed macros.
+
+## Prior Art
+
+The [Ember Decorators](http://ember-decorators.github.io/ember-decorators/) project has been experimenting with using decorators within Ember for some time now, with the goal of reaching feature parity with the classic object-model, and the learnings from that project will be used to inform the API design in this RFC.
+
+## Detailed design
+
+This RFC proposes that:
+
+1. `Ember.computed`, the `inject` macros, and the computed property macros be updated to return a standard javascript decorator. Internally, the classic object-model will be made aware of these decorators, and know how to apply them. This will allow the same exact imports to continue working as both native decorators and classic decorators.
+2. A new `@action` decorator be added to allow the definition of actions on classes.
+
+This does leave out some features from the classic model which are currently provided by Ember Decorators. Addressing them individually:
+
+* **Observers and event listeners** have long been considered an anti-pattern within Ember, and will eventually be removed from the framework. With the Octane model becoming the new standard, it doesn't make sense to continue supporting them with new APIs.
+* **Curly component** functionality such as `classNames`, `classNameBindings`, `attributeBindings`, etc. will be unnecessary with Glimmer components, and will in general be discouraged.
+* **Ember Data** provides computed properties which had to be manually wrapped in decorators. With the changes proposed in this RFC, however, they should continue to work without any additional changes. In fact, all computed property macros will.
+
+Moving forward, this RFC breaks down into _computed properties_ and _actions_.
+
+### Computed Properties
 
-For example, using the old `computed`:
+As mentioned before, computed properties essentially _are_ decorators. However, they are not spec compliant. Currently, `computed()` returns an instance of the `ComputedProperty` class, which contains all of the meta information about the decorated property. Native decorators, by contrast, are functions which receive a descriptor and modify it as necessary.
+
+Unfortunately, there's no way for us to know when a computed property is going to be used as a native decorator in a native class, and when it is going to be used in a classic class. This is especally true for computed macros, where arguments will be exactly the same in both cases:
```

Comment:
maybe I'm not understanding here, but this is some logic I've been in-progress on here: https://github.com/emberjs/ember.js/pull/17226/files#diff-aa29907c3a02fdab7b969fbf67e81c62R718

or do you maybe mean that decorators will be receiving the prototype in both scenarios?

Reply:
Yeah this is kind of hard to explain, which is why I added the example that follows it. Think about the following:

```js
class Foo {
  @alias('bar') baz;
}
```

Really, what's going on their is _not_ that we are invoking a decorator with parameters. We are invoking a function which _returns_ a decorator, so it desugars to:

```js
const aliasForBar = alias('bar');

class Foo {
  @aliasForBar baz;
}
```

Therefor, the `alias` function _must_ itself return a decorator function. However, this conflicts with usage in the classic programming model:

```js
const Foo = EmberObject.extend({
  baz: alias('bar')
})
```

We just established that `alias` must return a decorator function, but here it is with the exact same arguments, and it needs to return a `ComputedProperty` instance. There is nothing we can branch on here - in both cases, `alias` only receives the string `'bar'`, so it has no context for how it will be used.